zephyr/include/rtos/interrupt.h

```diff
@@ -0,0 +1,105 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ *
+ * TODO: IMX support, today IRQ is supported via SOF xtos API.
+ */
+
+#ifndef __ZEPHYR_RTOS_INTERRUPT_H__
+#define __ZEPHYR_RTOS_INTERRUPT_H__
+
+#if defined(CONFIG_IMX)
+/* imx currently has no IRQ driver in Zephyr so we force to xtos IRQ */
+#include "../../../xtos/include/rtos/interrupt.h"
+#else
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <stdint.h>
+
+/* to be removed once driver upstream in Zephyr */
+#define __SOF_DRIVERS_INTERRUPT_H__
+#include <platform/drivers/interrupt.h>
+
+static inline int interrupt_register(uint32_t irq, void(*handler)(void *arg), void *arg)
+{
+#ifdef CONFIG_DYNAMIC_INTERRUPTS
+	return arch_irq_connect_dynamic(irq, 0, (void (*)(const void *))handler,
+					arg, 0);
+#else
+	tr_err(&zephyr_tr, "Cannot register handler for IRQ %u: dynamic IRQs are disabled",
+		irq);
+	return -EOPNOTSUPP;
+#endif
+}
+
+/* unregister an IRQ handler - matches on IRQ number and data ptr */
+static inline void interrupt_unregister(uint32_t irq, const void *arg)
+{
+	/*
+	 * There is no "unregister" (or "disconnect") for
+	 * interrupts in Zephyr.
+	 */
+	irq_disable(irq);
+}
+
+/*
+ * CAVS IRQs are multilevel whereas BYT and BDW are DSP level only.
+ */
+static inline int interrupt_get_irq(unsigned int irq, const char *cascade)
+{
+#if CONFIG_SOC_SERIES_INTEL_ADSP_BAYTRAIL ||\
+	CONFIG_SOC_SERIES_INTEL_ADSP_BROADWELL || \
+	CONFIG_LIBRARY
+	return irq;
+#else
+	if (cascade == irq_name_level2)
+		return SOC_AGGREGATE_IRQ(irq, IRQ_NUM_EXT_LEVEL2);
+	if (cascade == irq_name_level5)
+		return SOC_AGGREGATE_IRQ(irq, IRQ_NUM_EXT_LEVEL5);
+
+	return SOC_AGGREGATE_IRQ(0, irq);
+#endif
+}
+
+/* enable an interrupt source - IRQ needs mapped to Zephyr,
+ * arg is used to match.
+ */
+static inline uint32_t interrupt_enable(uint32_t irq, void *arg)
+{
+	irq_enable(irq);
+
+	return 0;
+}
+
+/* disable interrupt */
+static inline uint32_t interrupt_disable(uint32_t irq, void *arg)
+{
+	irq_disable(irq);
+
+	return 0;
+}
+
+/* TODO: duplicate ? disables all IRQ sources on current core */
+#define arch_interrupt_disable_mask(x)	\
+	irq_lock()
```

Comment:
this should just disable interrupts from the mask, not all? And not at the DSP but at the interrupt controller?

Reply:
actually done this the other way around now and define irq_lock() for SOF users (and its used in 1 place with 0xffffffff).